service/history/api/pollupdate/api.go

```diff
@@ -66,34 +71,63 @@ func Invoke(
 			workflow.LockPriorityHigh,
 		)
 		if err != nil {
-			return nil, false, err
+			return nil, nil, false, err
 		}
 		release := wfctx.GetReleaseFn()
 		defer release(nil)
 		upd, found := wfctx.GetUpdateRegistry(ctx).Find(ctx, updateRef.UpdateId)
-		return upd, found, nil
+		wfKey := wfctx.GetWorkflowKey()
+		return &wfKey, upd, found, nil
 	}()
 	if err != nil {
 		return nil, err
 	}
 	if !ok {
 		return nil, serviceerror.NewNotFound(fmt.Sprintf("update %q not found", updateRef.GetUpdateId()))
 	}
-	namespaceID := namespace.ID(req.GetNamespaceId())
-	namespaceRegistry, err := shardContext.GetNamespaceRegistry().GetNamespaceByID(namespaceID)
-	if err != nil {
-		return nil, err
-	}
-	serverTimeout := shardContext.GetConfig().LongPollExpirationInterval(namespaceRegistry.Name().String())
-	// If the long-poll times out due to serverTimeout then return a non-error empty response.
-	stage, outcome, err := upd.WaitLifecycleStage(ctx, waitStage, serverTimeout)
-	if err != nil {
-		return nil, err
+
+	var stage enumspb.UpdateWorkflowExecutionLifecycleStage
+	var outcome *updatepb.Outcome
+
+	switch waitStage {
+	case unspecifiedStage:
+		stage, outcome, err = upd.LifecycleStage()
+		if err != nil {
+			return nil, err
+		}
+	case acceptedStage, completedStage:
+		stage, outcome, err = waitLifecycleStage(ctx, upd, waitStage, shardContext, req.GetNamespaceId())
+		if err != nil {
+			return nil, err
+		}
+	default:
+		return nil, serviceerror.NewInvalidArgument(fmt.Sprintf("support for LifecycleStage=%v is not implemented", waitStage))
 	}
 	return &historyservice.PollWorkflowExecutionUpdateResponse{
 		Response: &workflowservice.PollWorkflowExecutionUpdateResponse{
 			Outcome: outcome,
 			Stage:   stage,
+			UpdateRef: &updatepb.UpdateRef{
+				WorkflowExecution: &commonpb.WorkflowExecution{
+					WorkflowId: wfKey.WorkflowID,
+					RunId:      wfKey.RunID,
+				},
+				UpdateId: updateRef.UpdateId,
+			},
 		},
 	}, nil
 }
+
+func waitLifecycleStage(ctx context.Context, upd *update.Update,
+	waitStage enumspb.UpdateWorkflowExecutionLifecycleStage,
+	shardContext shard.Context, namespaceId string) (stage enumspb.UpdateWorkflowExecutionLifecycleStage, outcome *updatepb.Outcome, err error) {
+
+	namespaceID := namespace.ID(namespaceId)
+	namespaceRegistry, err := shardContext.GetNamespaceRegistry().GetNamespaceByID(namespaceID)
```

Comment:
Question: Is it registry or namespace. What does GetNamespaceByID return?

Reply:
Oh, good catch. This bad variable name was in the previous code, but I didn't notice it. It returns a namespace and I've changed the variable name accordingly.